// grunt-replace-url/Gruntfile.js
/*
 * grunt-marsrev
 * https://github.com/tdp100/grunt-marsrev
 *
 * Copyright (c) 2015 tangdeping
 * Licensed under the MIT license.
 */

'use strict';

module.exports = function (grunt) {
    // Project configuration.
    grunt.initConfig({
        jshint: {
            all: [
                'Gruntfile.js',
                'tasks/*.js',
                '<%= nodeunit.tests %>'
            ],
            options: {
                jshintrc: '.jshintrc'
            }
        },
        remoteResource: {
          staticResourceServer:"http://161.17.250.236/",
          frameResourceDir: "framework",
          frameworkVersion: "v2.1.5"
        },

        // Before generating any new files, remove any previously-created files.
        clean: {
            tests: ['build', 'tmp']
        },

        // Configuration to be run (and then tested).
        replace_url: {
          options: {
            mainJs: {
              filePath: 'build/main.js',   //data-main 文件的路径
              configFile: 'main.json',
              remotePaths: ["can", "app-remote", "ui-router", "bootstrap", "moment", "language-remote", "lazy-load"]
            },
            cwd: 'build',                       //必填, 要hash的所有文件的目录名
            //targetFiles: [{src: ['index.html']}],
            targetFiles: [{src: ['/']}],
            files: [
              {src: ['/', '!i18n/en']}
              //{src: ['./theme/default/images/favicon.icon', './theme/default/css/combined-framework.min.css',
              //'./urlDeal.js', './browserChecker.js', './lib/require.js', './lib/tiny/tiny-1.34.0-rc1-min.js', 'main.js']}
            ]
          }
        }
    });

    // Actually load this plugin's task(s).
    grunt.loadTasks('tasks');

    // These plugins provide necessary tasks.
    grunt.loadNpmTasks('grunt-contrib-jshint');
    grunt.loadNpmTasks('grunt-contrib-clean');

    // Whenever the "test" task is run, first clean the "tmp" dir, then run this
    // plugin's task(s), then test the result.
    grunt.registerTask('test', ['clean', 'replace_url']);

    // By default, lint and run all tests.
    grunt.registerTask('default', ['jshint', 'test']);

};
